common/src/test/java/com/microsoft/identity/common/MicrosoftFamilyOAuth2TokenCacheTest.java

```diff
@@ -525,4 +534,146 @@ public void testDualStackIdTokenRetrieval() throws ClientException {
+                    frtTestBundle2.mGeneratedAccessToken.setSecret(SECRET + "AT2" + currentThread);
+                    return frtTestBundle2.mGeneratedAccessToken;
+                }
+                return null;
+            }
+        });
+        when(mockCredentialAdapter.createRefreshToken(mockStrategy, mockRequest, mockResponse)).thenAnswer(new Answer<RefreshTokenRecord>() {
+            @Override
+            public RefreshTokenRecord answer(InvocationOnMock invocation) throws Throwable {
+                String currentThread = Thread.currentThread().getName();
+                if(currentThread == thread1[0]) {
+                    frtTestBundle1.mGeneratedRefreshToken.setSecret(SECRET + "RT2" + Thread.currentThread().getName());
+                    return frtTestBundle1.mGeneratedRefreshToken;
+                } else if(currentThread == thread2[0]) {
+                    frtTestBundle2.mGeneratedRefreshToken.setSecret(SECRET + "RT2" + Thread.currentThread().getName());
+                    return frtTestBundle2.mGeneratedRefreshToken;
+                }
+                return null;
+            }
+        });
+        when(mockCredentialAdapter.createIdToken(mockStrategy, mockRequest, mockResponse)).thenAnswer(new Answer<IdTokenRecord>() {
+            @Override
+            public IdTokenRecord answer(InvocationOnMock invocation) throws Throwable {
+                String currentThread = Thread.currentThread().getName();
+                if(currentThread == thread1[0]) {
+                    return frtTestBundle1.mGeneratedIdToken;
+                } else if(currentThread == thread2[0]) {
+                    return frtTestBundle2.mGeneratedIdToken;
+                }
+                return null;
+            }
+        });
+
+        // Submit runnable tasks to save tokens from two different threads concurrently
+        Runnable runnable1 = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    thread1[0] = Thread.currentThread().getName();
+                    mOauth2TokenCache.save(mockStrategy, mockRequest, mockResponse);
+                } catch (ClientException e) {
+                    throw new AssertionError(e.getMessage());
+                }
+            }
+        };
+
+        Runnable runnable2 = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    thread2[0] = Thread.currentThread().getName();
+                    mOauth2TokenCache.save(mockStrategy, mockRequest, mockResponse);
+                } catch (ClientException e) {
+                    throw new AssertionError(e.getMessage());
+                }
+            }
+        };
+
+        ExecutorService executorService = Executors.newFixedThreadPool(5);
+        Future runnableTask1 = executorService.submit(runnable1);
+        Future runnableTask2 = executorService.submit(runnable2);
```

Comment:
There's no guarantee that these are actually not happening sequentially here.  You'll want to put a latch in place to insure that we have problems.

Reply:
done